File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int i;
    private int p;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return builder(preorder, inorder, Integer.MIN_VALUE);
    }

    private TreeNode builder(int[] preorder, int[] inorder, int stop) {
        // preorder 유효성 체크 메소드
        // inorder 유효성 체크 메소드
        if (p >= preorder.length) return null;
        if (inorder[i] == stop) {
            i += 1;
            return null;
        }

        TreeNode node = new TreeNode(preorder[p]);
        p += 1;

        node.left = builder(preorder, inorder, node.val);
        node.right = builder(preorder, inorder, stop);
        return node;
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        self.i = 0  # inorder 배열의 인덱스
        self.p = 0  # preorder 배열의 인덱스

        def builder(stop):
            if self.p >= len(preorder):
                return None
            if inorder[self.i] == stop:
                self.i += 1
                return None

            # 새로운 노드 생성
            node = TreeNode(preorder[self.p])
            self.p += 1

            # 왼쪽 서브트리 생성
            node.left = builder(node.val)
            # 오른쪽 서브트리 생성
            node.right = builder(stop)

            return node

        # 트리 생성 시작
        return builder(float('-inf'))